WIP [SDK] - Proposal to split Helm/Ansible on their repos (#44)
operator-framework/enhancements: enhancements/split_helm_ansible_into_their_repos.md

---
title: Split Helm/Ansible into their own repos
authors:
  - "@camilamacedo"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2020-08-26
last-updated: 2020-08-30
status: implementable
see-also:
  - "/enhancements/this-other-neat-thing.md"  
replaces:
  - "/enhancements/that-less-than-great-idea.md"
superseded-by:
  - "/enhancements/our-past-effort.md"
---

# Split Helm/Ansible into their projects

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [x] Test plan is defined
- [x] Graduation criteria for dev preview, tech preview, GA

## Open Questions 

For Helm, besides the requirement of split the repo it also has a new pkg implementation which requires to be putted in place and is in the [joelanford/helm-operator](https://github.com/joelanford/helm-operator) currently. 

This proposal is decoupling both requirements, which allow a specific strategy for each one of them. However, would this option be more comfortable to move forward with both needs or the approach to address them for Helm simultaneously could be more practical?

> **Note** A enhancement to describe the possible options to address the Helm requirement for the new pkg implementation ought to be made.

Author: It was done. See: https://github.com/operator-framework/enhancements/pull/45/files
If that gets merged than we can update this doc with a link for this proposal.

## Summary

This proposal has the goal to describe how we can split Helm/Ansible projects and then, decoupled from the CLI tool. 

**Note** It is a follow up of the proposal [Split SDK into multiple projects](split-sdk-into-multiple-projects.md).   

## Motivation

- Improve the maintainability and reusability by adopting concepts such as [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) and [Single Responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). 
- Have healthier lifecycle since separately and individually, which means that we can have a single/granular releases.

> - Solve the most significant part of the problems faced with the CI. Currently, SDK repo takes to long to pass in the tests, which indeed is not just a problem to contribute to the projects but for us be able to cover more scenarios and ensure quality. 

Author: I replaced it by `Minimize the CI processing time`. however, I think it is also a motivation for us to put it in place.

> - Ensure that any type/language operator will be able to be built following the same base config and using a common lib which will be helpful to keep them maintained and supported as to provide more types of operators such as hybrids and done for other languages.  

Author: I provided more details on the doc but yes. the config/ stuff and IMO we can also add the Makefile Bollerplate as well.

### Goals

> - The SDK CLI tool repository has just the code implementation, which is part of its domain of responsibility. 

Reviewer: ```suggestion
- The `operator-framework/operator-sdk` repository only contains CLI and Go plugin code.
```

Author: The operator-sdk do not contain the Go plugin. Its implementation is on Kubebuilder. Just the call/usage is done in SDK.

> - A Helm Operator repository with all code implementation which is specific to it and is part of its domain of responsibility

Reviewer: ```suggestion
- The `operator-framework/helm-operator` repository only contains `helm-operator` code and Helm plugin code.
```

> - An Ansible Operator repository with all code implementation which is specific to it and is part of its domain of responsibility

Reviewer: ```suggestion
- The `operator-framework/ansible-operator` repository only contains `ansible-operator` code and Ansible plugin code.
```

> - An SDK Operator Lib which contains the standard code that should be used to generate all types of Operators. (e.g. utils for releases which might be expected for all, templates and the implementation of the plugins which can be used for any language/type of operator)

Reviewer: ```suggestion
- Code shared by `operator-sdk`, `helm-operator`, and `ansible-operator` repos will be placed in the `operator-framework/operator-lib` repository when possible.
```

Author: operator-sdk should not use the operator-build-lib since it should no longer have code to build the projects and will just use the plugins for that. But the text was not good here. I changed let me know if it is better now.

> - Ansible and Helm repository have their godocs generated automatically 

Reviewer: This is always the case with godocs.
```suggestion
```

Author: I changed for `- Ansible/Helm/SDK Operator Build Lib repositories with lint check to ensure that all expose methods will be documented`. I think it clarifies better. Thank you.

- Ansible and Helm repository have their CI, makefile targets and helpers and utilities required to contribute to them individually and test the projects scaffolded for these types. 
- Helm/Ansible binaries should be created/released from their repos and no longer from SDK CLI. 

### Non-Goals

> - Apply the changes for new Helm based-operator implementation in the [joelanford/helm-operator](https://github.com/joelanford/helm-operator) repository.

Reviewer: Is there a good reason to mention this in this EP? Is @joelanford's `helm-operator` implementation discussed in the SDK's docs currently?

Author: I think you are right @estroz. I removed it from here and just leave in the Alternatives 👍

- Steps required to achieve the goal of the types projects have their own docs subdomains `ansible.operatorframework.io` and `helm.operatorframework.io`  
 
## Proposal

> ### I am as a mantainer and/or community contributor would like to 

Reviewer: Organize these into user stories.

> - be able to check the CI make its work in a reasonable time

Reviewer: This is not a user story but is a nice side effect of this change. See above.

- be able to quickly identify what part of the domain of responsibility of each type is and what is shared between them. 

> - be able to create new Operator types using the SDK Operator lib easily

Reviewer: Do you mean new project types? If so this is definitely out of scope of this EP.

Author: By `Create a new repository SDK Operator Build Lib` we can address the following needs automatically. However, I will clarify that in the NO Goals as well.  👍 

- I am a maintainer/contributor, I want to be able to provide helpers for users to create your own types, so that I can easily join these new types in the project incubator
- I am a maintainer/contributor, I want to be able to provide helpers for users create your own types following that same standards and workflows, so that I can easily ensure that its types will be integrated with OLM and will work well with the SDK features

- be able to easily maintain new Operator types by updating its lib and not caring about changes made in the common base.
- be able to patch/update just one place when we bump new upstream versions instead of needing to update its duplications across all projects.   
- be able to provide releases according to the lifecycle of each type individually

> ### I am as a user would like to 

Reviewer: Organize these into user stories.

- upgrade/migrate my projects just only if a new release contains changes for the specific type of the used project 

> - develop my own types base on the helpers provided by the SDK Operator lib 

Reviewer: Do you mean new project types? If so this is definitely out of scope of this EP.

Author: Clarified

- have better visibility about what was changed in the SDK releases and per type

> - not have more than one standard/command or workflow to achieve the same need with different types of project

Reviewer: I don't see how this relates to the changes proposed here. `operator-sdk` user workflows aren't changing.

Author: By centralizing the common code in the operator-build-lib we can better ensure that all types will use the same base. For example. My idea here we could provide a base for the files scaffolds on the `config/`. Also, the code in internal/plugins to customize the Makefile would be also centralized so it would ensure that any operator build using it would have the same targets.

> - not need to learn how to configure and working on with the operator projects built with SDK CLI tool from scratch when I need to create operators with different types.   

Reviewer: What do you mean by this?

Author: I fixed the info here. See: 

- I am an operator developer, I want to be able to create my own operator types, so that I can attend my needs even when the maintainers of operator-sdk do not consider it viable to support the type I need
- I am an operator developer, I want to that all types of operators following the same standard and have the same workflow, so that I do need lear all from scratch when I try to attend my need with other languages
- I am an operator developer, I want to that all common commands and Makefile helpers are the same for any type, so that I will not be confused when I use another type of operator
- I am an operator developer, I want to that all operators types are configured at the same way, so that I can easily work with more than one type and by learning how one works I will know how all of them works as well.

Let me know if it is more clear now.

- use the Helm/Ansible pkg implementation to create my customized helm/ansible based-operator images 

### Implementation Details/Notes/Constraints 

> - The code implementation which will be moved from SDK repository to the new types repositories needs to keep their GitHub history tracked. 

Reviewer: This is a pretty significant constraint. Can you expand upon how this will be done?

Author: I thought that @jmrodri could find a solution to do it for operator-lib. However, I just checked its logs and shows that not. Removing it 👍

- The SDK repository will require cleanup in its e2e tests and makefile targets

### Risks and Mitigations

The most significant risk is the instability that will occur while the projects are split apart. Once all the projects are in place, things will come back together.

## Design Details

### Operator Type Projects 

#### Layout

The following shows the final `helm-operator` layout with the description from where its implementation came from. See that `ansible-operator` will be very similar and has the same structure in the SDK repository which means that the following information is valid for both. 

> ```shell
> .

Reviewer: ````suggestion
```
````

    ├── README.md 
    ├── go.mod
    ├── go.sum
    ├── main.go (used to run the project locally. Move operator-sdk/internal/cmd/helm-operator/run from SDK to here)
    ├── hack (store the shell scripts)
         └── test (contains the scripts required to run the e2e tests)        
    ├── pkg 
         └── helm (contains all based-operator image implementation and helpers. Move operator-sdk/internal/helm/ from SDK to here)        
    |   └── plugin 
            └── v1 (we will move operator-sdk/internal/plugins/helm/v1/ from SDK to here)
    ├── testdata
    │   └── test-chart-0.1.0.tgz (mock data to test the project if required)
    │   └── example (sample project that should be automatically generated based on the pr changes)
    ├── e2e-tests (should contain the e2e test of the projects. Note that it means copy operator-sdk/test/e2e-helm from SDK to here and cleanup the SDK test implementation since it will no longer required to test/ensure the Helm features)
    └── version
        └── version.go (Move operator-sdk/cmd/helm-operator/version from SDK to here)
```

#### Helm

> | From SDK repo| To Helm repo	| Description  	|
> |---	|---	|---	|
> |  operator-sdk/internal/cmd/helm-operator/run 	|   helm-operator/main.go	|   used to run the project locally	|

Reviewer: Remove the repo prefixes, it makes reading this table easier. Ex:
```suggestion
| From `operator-sdk`| To `helm-operator`	| Description  	|
|---	|---	|---	|
|  internal/cmd/helm-operator/run 	|   main.go	|   helm-operator's `main` package	|
```

| operator-sdk/internal/cmd/helm-operator/version  	|  helm-operator/version.go 	| define the version of the project and it used to build the bin	|
|  operator-sdk/internal/helm/ 	|  helm-operator/pkg/ 	|   has the implementations used to generated the Helm based-operator image	|
|  operator-sdk/internal/plugins/helm/v1/ 	|  helm-operator/pkg/plugin/v1 	|   has the implementation customized to scaffold the helm projects 	|
|  operator-sdk/test/e2e-helm 	|  helm-operator/e2e-tests 	|   has the e2e test for the Helm project 	|
|  operator-sdk/hack/tests/e2e-helm.sh	|  helm-operator/hack/tests/e2e-helm.sh 	|   has the e2e test for the Helm project 	|

#### Ansible

> | From SDK repo| To Ansible repo	| Description  	|

Reviewer: Remove the repo prefixes, it makes reading this table easier.

**NOTE**: The molecule tests still via shell script, however, if they be migrated to go test stack then the mock static files will no longer be required. 

### SDK Operator lib

The idea of this project as described above is to centralize all code which is useful for all types projects. In this way, SDK CLI as any operator type repo can import this lib and use its features. 
 
In this way follow up the SDK tree with the points that would be moved:

- `internal/util/k8sutil/constants.go`: Just the `WatchNamespaceEnvVar` should be moved since it is used for both types in the `run.go` implementation and not used by the SDK CLI commands at all. 
- `internal/plugins`: code implementation which is used to customize the scaffold of the projects and allow they work on and be integrated with the framework features (plugins phase 2). Note that ansible/helm/go plugins will not be moved to this repository.    
- `test/internal`: the utils tests implemented in SDK can will used for any operator type and SDK CLI repo as well to implement the e2e tests.  

### Usage of Helm/Ansible plugin in SDK repo 

> Their usage in the SDK repo will be done as it currently is made for the Go operator. See: https://github.com/operator-framework/operator-sdk/tree/master/internal/plugins/golang/v2

Reviewer: You just linked the plugin itself which isn't as helpful as the `CLI` usage:
```suggestion
Helm and Ansible plugins will be imported and added to `operator-sdk`'s CLI much like they [currently are](https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L62-L63), except from their respective repositories.
```

## Test Plan

> The ultimate test is that e2e tests and manual testing of the SDK binary will

Reviewer: We shouldn't be talking about manual testing here (it's a throwaway comment since we're not talking about it elsewhere), and e2e tests should be discussed only in the context of individual plugins and operators. We need to discuss how _integration_ tests will be used to determine correctness of the `operator-sdk` binary.

Author: I am closing this one because I better specified how I am thinking about the tests. However, please, feel free to check and give your inputs.

be used to determine it remains functional. Each repo will contain their own e2e and unit tests which ensure that they all will still working as should be after the split.
 
Besides Ansible/Helm be decoupled from SDK CLI tool their tests should still ensure their functionalities with the tool which means that we will need to test them with the SDK binary provide from master. 

Also, to avoid the risk and scenarios where the changes in the tool can impact in their behaviours and functionalities still required have in SDK e2e test which will scaffold the projects for each type and test the tool integration  with them such as to ensure scorecard and OLM features. 

In this way we will have:

- Ansible/Helm repositories with unit and e2e tests 
- SDK repository with the pieces of the e2e tests which ensure its features
- SDK Operator lib covered with unit tests and tested indirectly as well via the SDK/ Ansible/Helm repo since the lib will be imported and used by them.  

## Examples

> The project [joelanford/helm-operator](https://github.com/joelanford/helm-operator) is an excellent example to illustrate and demonstrate its motivations and the expected final result. Note that it is not using the SDK Operator lib as other granular details described here.   

Reviewer: We say that using this operator is a non-goal, but then reference it in examples? I'd rather not talk about it at all in that case.

**Note** It is a follow up of the proposal [Split SDK into multiple projects](split-sdk-into-multiple-projects.md).   

## Open Questions

1. Any reason for not generate the samples on the Operator-Type repos? 
1. The templates copied from Kubebuilder were set in Kubebuilder as internal by us in the plugin phase 1 implementation. We use the plugins, and the templates are part of them, so I do not see any reason for we not be allowed to import its templates as well. However, do you see any objection to it?      
1. The steps to reach the goal for the Operator Types projects have their own docs subdomains `ansible.operatorframework.io`, and `helm.operatorframework.io` shows have certain complexity. Have we any reason for it has not its specific proposal?

## Motivation

- Improve the maintainability, readability and reusability by adopting concepts such as [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) and [Single Responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). 
- Facilitate the development process and encourage more contributions
- Allow Lifecycle releases per type

> - Minimize the CI processing time

Author: Hi @estroz, IMO the time to process the CI has been a big problem for us. So, I agree with you that it is not a Goal. However, I think that it is a Motivation for we put it in place.

This solution could be helpful to provide alternatives to developers quickly built their Operator Types such as to support [Argo](https://github.com/operator-framework/operator-sdk/issues/2497) and to allow the common needs to be re-used and be easily support new Operator Types such as [Hybrid Operators](https://github.com/operator-framework/operator-sdk/issues/670) or with other languages see [Meta: Operator SDK's in other languages #2745](https://github.com/operator-framework/operator-sdk/issues/2745), so that we could easily join these new Operator Types as incubator projects and define as pre-requirement following the same standards and workflows to work with the SDK features. However, it is definitely out of the scope of this EP. 

> 

Author: Hi @estroz, 

I tried to clarify my ideas here. Besides these needs are not a Go of this EP I think it is important we have them in mind since they are ultimate goals for the projects. Otherwise, we might take decisions that will be harder to reach them in the future.

Note that, the above scenarios requires small changes in the docs usually. Also, changes in the tutorials will be more common when the upstream (Kubebuilder) implementation change and this case, also will be required update the website docs when a new Kubebuilder version be bumped. So, it shows achievable. 
  
### Operator Types Samples 

See the proposal [Automatic Samples Generation using Go language](https://github.com/operator-framework/enhancements/pull/47/files). The implementation to generate the samples (SampleContext) as it helpers could be on upstream. 
 
Then, each Operator Type should be able to build its own samples in the `testdata/` directory.

### Scope of Tests per project

- Operator Types (`helm-operator` and `ansible-operator`) should contains tests that will cover its libs and the project scaffold by them.  

- SDK CLI tool (`operator-sdk`) should contain the test required to cover the CLI commands and e2e tests which will cover the specific customizations made on top of helm/ansible plugins for example olm integrations. 

> 

Author: HI @estroz, 

I tried to put here what in my mind should be covered on the tests per project.

- I am an operator developer, I want to see all Operator Types following the same standard configuration when possible, so that I switch between the available types to address more appropriated my needs with less complexity.

#### Split Helm/Ansible into their own repos

- I am a maintainer/contributor, I want to see the projects adopting concepts such as [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) and [Single Responsibility principle](https://en.wikipedia.org/wiki/Single, so that I can maintain, read and re-use them easily
- I am a Helm/Ansible operator developer, I want to need to update my projects only when the new release address changes my type of project, so that I can easily keep my operator projects updated. 

> - I am a Helm/Ansible operator developer, I want to need to update my projects only when the new release address changes my type of project, so that I can easily keep my operator projects updated. 

Reviewer 2: Nit: duplicate
